src/crimson/osd/pg_map.cc

```diff
@@ -19,75 +19,134 @@ seastar::future<core_id_t> PGShardMapping::get_or_create_pg_mapping(
   } else {
+    DEBUG("calling primary to add mapping for pg {} to the expected core {}",
+          pgid, core_expected);
     return container().invoke_on(
-        0, [pgid, core_expected, FNAME](auto &primary_mapping) {
-      auto [insert_iter, inserted] =
-        primary_mapping.pg_to_core.emplace(pgid, core_expected);
-      ceph_assert_always(inserted);
-      ceph_assert_always(primary_mapping.core_to_num_pgs.size() > 0);
-      std::map<core_id_t, unsigned>::iterator core_iter;
-      if (core_expected == NULL_CORE) {
-        core_iter = std::min_element(
-          primary_mapping.core_to_num_pgs.begin(),
-          primary_mapping.core_to_num_pgs.end(),
-          [](const auto &left, const auto &right) {
-            return left.second < right.second;
+        0, [pgid, _core_expected=core_expected, FNAME](auto &primary_mapping) {
+      auto core_expected = _core_expected;
+      auto find_iter = primary_mapping.pg_to_core.find(pgid);
+      if (find_iter != primary_mapping.pg_to_core.end()) {
+        auto core_found = find_iter->second;
+        assert(core_found != NULL_CORE);
+        if (core_expected != NULL_CORE) {
+          if (core_expected != core_found) {
+            ERROR("the mapping is inconsistent for pg {} (primary): core {}, expected {}",
+                  pgid, core_found, core_expected);
+            ceph_abort("The pg mapping is inconsistent!");
           }
-        );
-        core_expected = core_iter->first;
-      } else {
-        core_iter = primary_mapping.core_to_num_pgs.find(core_expected);
+          // core_expected == core_found
+          DEBUG("mapping pg {} to core {} (primary): already mapped and expected",
+                pgid, core_expected);
+        } else { // core_expected == NULL_CORE
+          core_expected = core_found;
+          DEBUG("mapping pg {} to core {} (primary): already mapped",
+                pgid, core_expected);
+        }
+        // proceed to broadcast core_expected
+      } else { // find_iter == primary_mapping.pg_to_core.end()
+        ceph_assert_always(primary_mapping.core_to_num_pgs.size() > 0);
+        std::map<core_id_t, unsigned>::iterator count_iter;
+        if (core_expected == NULL_CORE) {
+          count_iter = std::min_element(
+            primary_mapping.core_to_num_pgs.begin(),
+            primary_mapping.core_to_num_pgs.end(),
+            [](const auto &left, const auto &right) {
+              return left.second < right.second;
+            }
+          );
+          core_expected = count_iter->first;
+        } else { // core_expected != NULL_CORE
+          count_iter = primary_mapping.core_to_num_pgs.find(core_expected);
+        }
+        ceph_assert_always(primary_mapping.core_to_num_pgs.end() != count_iter);
+        ++(count_iter->second);
+        auto [insert_iter, inserted] =
+          primary_mapping.pg_to_core.emplace(pgid, core_expected);
+        assert(inserted);
```

Comment:
The fix here is to emplace the new mapping **only** if it wasn't already mapped, right? (under `find_iter == primary_mapping.pg_to_core.end()`)

Reply:
> @cyx1231st it is also create pg on shard 0 and broadcast to others, why you name the commit : allow multiple shards to create new pg mappings at the same time

@liu-chunmei

Reply:
Right, because `primary_mapping.core_to_num_pgs` also needs to keep correct.